src/lib.rs

```diff
@@ -0,0 +1,189 @@
+        let actual_token_json = json::parse(&access_token).unwrap();
+        let actual_token = actual_token_json["access_token"].as_str().unwrap();
+        let member = rt.block_on(get_guild_member(actual_token, guild_id)).unwrap();
+        let member_json = json::parse(&member).unwrap();
+        if member_json["roles"].contains(role){
+            return PamResultCode::PAM_SUCCESS
+        }
+        pam_try!(conv.send(PAM_ERROR_MSG, "You do not have the required role!"));
+        PamResultCode::PAM_PERM_DENIED
+    }
+
+    fn sm_setcred(_pamh: &mut PamHandle, _args: Vec<&CStr>, _flags: PamFlag) -> PamResultCode {
+        println!("set credentials");
+        PamResultCode::PAM_SUCCESS
+    }
+
+    fn acct_mgmt(_pamh: &mut PamHandle, _args: Vec<&CStr>, _flags: PamFlag) -> PamResultCode {
+        println!("account management");
+        PamResultCode::PAM_SUCCESS
+    }
+}
+
+async fn generate_new_token(state: &str, port: u16) -> String {
+    let addr = ([0, 0, 0, 0], port).into();
+
+    let notify = Arc::new(Notify::new());
+    let token = Arc::new(Mutex::new(None::<String>));
+    let make_svc = make_service_fn({
+        let notify = notify.clone();
+        let token = token.clone();
+        move |_| {
+            let notify = notify.clone();
+            let token = token.clone();
+            async move {
+                Ok::<_, Infallible>(service_fn(move |req: Request<Body>| {
+                    let notify = notify.clone();
+                    let token = token.clone();
+                    async move {
+                        let code = req
+                          .uri()
+                          .query()
+                          .map(|query|
+                            url::form_urlencoded::parse(query.as_bytes())
+                          )
+                          .and_then(|mut query|
+                            query.find(|(name, _)| name.to_string() == "code")
+                          )
+                          .map(|(_, code)| code.into_owned());
+                        if let (Some(code), token @ None) = (code, &mut *token.lock().await) {
+                          notify.notify_one();
+                          *token = Some(code);
+                        }
+                        Ok::<_, Infallible>(Response::new(Body::from(
+                            "fuck you",
+                        )))
+                    }
+                }))
+            }
+        }
+    });
+    let server = hyper::Server::bind(&addr)
+        .serve(make_svc)
+        .with_graceful_shutdown(async move { tokio::select! {
+          _ = notify.notified() => (),
+          _ = sleep(Duration::from_secs(60 * 1)) => (),
+        }});
+    server.await.unwrap();
+
+    let mut lock = token.lock().await;
+    lock.take().unwrap()
+}
+
+async fn get_access_token(client_id: &str, client_secret: &str, code: String, url: String) -> Result<std::string::String, reqwest::Error> {
```

Comment:
When you are just reading strings, you should use `&str`. Using `String` here means that we are taking ownership of the string from the callee, which means they can't use the value of `code` after calling this function. This can result in uneeded clones.